### DISCUSSION TITLE
Audit feature in MongoDB Panache Quarkus (similar to spring @CreatedDate and @LastModifiedDate)

### DISCUSSION THREAD

**Discussion by nitinty** [quarkusio/quarkus#41612] [Q&A]
I'm working on a project where we need to migrate it from spring to Quarkus. We are trying to use MongoDB Panache.

We want to add the auditing while saving/updating the entity by using createdAt/modifiedAt in collection. Do we have anything similar to @CreatedDate and @LastModifiedDate like we have in Spring.

[a comment by an automated account is not shown]

**Reply by nitinty** [reply]
@FroMage  @geoand @loicmathieu 
Can you please assist here ?

**Reply by geoand** [reply]
We don't have that AFAIR. Also cc @loicmathieu

**Reply by FroMage** [reply]
I was going to comment that it would be nice in ORM/HR Panache too, because I've always used this pattern:

```java
	public Date lastUpdated;

	@PreUpdate
	@PrePersist
	public void prePersist() {
		lastUpdated = Date.from(Instant.now());
	}
```

But, apparently, there's `@CreationTimestamp` and `@UpdateTimestamp` already. I guess I should move to that!

**Reply by nitinty** [reply]
@FroMage 

Seems @CreationTimestamp and @UpdateTimestamp both are part of quarkus-hibernate-reactive-panache, Not as part of quarkus-mongodb-panache !!!

**Reply by FroMage** [reply]
I know, I didn't mean to imply that it worked for Mongo. Just that the ORM/HR side has it, so there's precedent. If we add it to Mongo/Panache, it may make sense to use the same names.

**Reply by nitinty** [reply]
Understood, Is there any plan to add it to Mongo/Panache in upcoming upgrade ?

**Reply by FroMage** [reply]
Not that I'm aware of, but if you feel like making a PR, we can help you :)

**Reply by loicmathieu** [reply]
`@CreationTimestamp` and `@UpdateTimestamp` are Hibernate specific not Quarkus specific, so if we want to do the same for MongoDB we cannot reuse the annotation and must implement what Hibernate implemented. 

By the way, you can implement it easily with a MongoDB command listener.

**Reply by nitinty** [reply]
@loicmathieu 


Yes I tried by implementing the **com.mongodb.event.CommandListener**, But the problem is we are not getting the type of Entity in the **commandStarted(CommandStartedEvent event)** method to check whether that particular entity has the mentioned annotation on any field. And I want to avoid additional config for the collection_name to Entity_type.

**Reply by loicmathieu** [reply]
If you use the repository pattern, you can add an annotation to the method making the update and use a CDI interceptor instead.

**Reply by nitinty** [reply]
I hope your are taking about the below approach, By using **@UpdateTimestamp**  and bind it with interceptor where we can update the timestamp. But here is the point, Is it good approach to initialize the entity level property, we have to use method level annotation ? (As interceptor not invoked with field level annotation). 


@ApplicationScoped
public class TestService {

    @Inject
    TestRepository testRepository;

    **@UpdateTimestamp**
    public void persistEntity(TestEntity entity) {
        testRepository.persist(entity);
    }
}

**Reply by loicmathieu** [reply]
Yes, you get the idea.

**Reply by nitinty** [reply]
@loicmathieu  

There is one more disadvantage with this repo pattern I have to define every method on RepoImpl even though  I can use this method directly by repo object. Can you please suggest me any other approach  ? Or is there any way to initialize it before saving/updating entity via any EntityCallback like we have in spring **org.springframework.data.mapping.callback.EntityCallback**

please assist.



```

@ApplicationScoped
public class UserRepositoryImpl implements UserRepository{
    @Override
    **@UpdateTimestamp**
    public void persistEntity(UserEntity userEntity) {
        UserRepository.super.persist(userEntity);
    }
}

```

**Reply by loicmathieu** [reply]
I cannot see any other way.
If you have multiple repositories, you can create an abstract repository that does that.
